src/t8_geometry/t8_geometry_implementations/t8_geometry_lagrange.cxx

```diff
@@ -0,0 +1,170 @@
+  for (int i_component = 0; i_component < T8_ECLASS_MAX_DIM; i_component++) {
+    double inner_product = 0;
+    for (int j_vertex = 0; j_vertex < n_vertex; j_vertex++) {
+      const double coordinate = active_tree_vertices[j_vertex * T8_ECLASS_MAX_DIM + i_component];
+      const double basis_function = basis_functions[j_vertex];
+      inner_product += basis_function * coordinate;
+    }
+    mapped_point[i_component] = inner_product;
+  }
+}
+
+const std::vector<double>
+t8_geometry_lagrange::compute_basis (const double *ref_coords) const
+{
+  switch (active_tree_class) {
+  case T8_ECLASS_TRIANGLE:
+    switch (*degree) {
+    case 2:
+      return t8_geometry_lagrange::t6_basis (ref_coords);
+    }
+  case T8_ECLASS_QUAD:
+    switch (*degree) {
+    case 1:
+      return t8_geometry_lagrange::q4_basis (ref_coords);
+    }
+  default:
+    SC_ABORTF ("Error: Lagrange geometry for degree %i %s not yet implemented. \n", *degree,
+               t8_eclass_to_string[active_tree_class]);
+  }
+}
+
+const std::vector<double>
+t8_geometry_lagrange::t3_basis (const double *ref_point) const
+{
+  const double xi = ref_point[0];
+  const double eta = ref_point[1];
+  const std::vector<double> basis_functions = { 1 - xi, xi - eta, eta };
+  return basis_functions;
+}
+
+const std::vector<double>
+t8_geometry_lagrange::t6_basis (const double *ref_point) const
+{
+  const double xi = ref_point[0];
+  const double eta = ref_point[1];
+  const std::vector<double> basis_functions
+    = { 1 - 3 * xi + 2 * xi * xi,      -xi + eta + 2 * xi * xi + 2 * eta * eta - 4 * xi * eta,
+        -eta + 2 * eta * eta,          4 * xi - 4 * eta - 4 * xi * xi + 4 * xi * eta,
+        -4 * eta * eta + 4 * xi * eta, 4 * eta - 4 * xi * eta };
+  return basis_functions;
+}
+
+const std::vector<double>
+t8_geometry_lagrange::q4_basis (const double *ref_point) const
+{
+  const double xi = ref_point[0];
+  const double eta = ref_point[1];
+  const std::vector<double> basis_functions = { (1 - xi) * (1 - eta), xi * (1 - eta), eta * (1 - xi), xi * eta };
+  return basis_functions;
+};
+
+T8_EXTERN_C_BEGIN ();
+
+/* Satisfy the C interface from t8_geometry_lagrange.h.
+ * Create a new geometry with given dimension. */
+t8_geometry_c *
+t8_geometry_lagrange_new (int dimension)
+{
+  t8_geometry_lagrange *geom = new t8_geometry_lagrange (dimension);
+  return (t8_geometry_c *) geom;
+}
+
+void
+t8_geometry_lagrange_destroy (t8_geometry_c **geom)
+{
+  T8_ASSERT (geom != NULL);
+  T8_ASSERT (t8_geom_is_lagrange (*geom));
+
+  delete *geom;
+  *geom = NULL;
+}
+
+#if T8_ENABLE_DEBUG
+int
+t8_geom_is_lagrange (const t8_geometry_c *geometry)
+{
+  /* Try to dynamic cast the geometry into Lagrange geometry. This is only successful if
+   * geometry pointed to a t8_geometry_lagrange.
+   * If successful, then is_lagrange_geom will be true.
+   */
+  const int is_lagrange_geom = (dynamic_cast<const t8_geometry_lagrange *> (geometry) != NULL);
+
+  return is_lagrange_geom;
+}
+#endif
```

Comment:
```suggestion
```
We have reworked this function. You can delete this here and I will write a further comment in you hxx file